### PR TITLE
Prepare to support 8ULP

### PR DESCRIPTION
**Pull request by TangleZ**
8ULP have no irqstr then have some change with interrupt module.

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test this please** to run this test once.

File: src/drivers/imx/irqstr.c [hunk 1]
```diff
@@ -97,7 +98,7 @@ DECLARE_TR_CTX(irq_i_tr, SOF_UUID(irq_imx_uuid), LOG_LEVEL_INFO);
 
 /* HW register access helper methods */
 
-static inline void irqstr_write(uint32_t reg, uint32_t value)
+void irqstr_write(uint32_t reg, uint32_t value)
```
Comment (paulstelian97) [inline src/drivers/imx/irqstr.c:101]: This is an auxiliary function, it should remain static inline.

File: src/drivers/imx/interrupt-irqstr.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	 */
+	irqstr_disable_hw();
+	/* Mask every external IRQ first */
+	for (i = 0; i < IRQSTR_IRQS_REGISTERS_NUM; i++)
+		irqstr_write(IRQSTR_CH_MASK(i), 0);
+	/* Turn on the IRQ_STEER hardware */
+	irqstr_enable_hw();
+
+	for (i = 0; i < ARRAY_SIZE(dsp_irq); i++)
+		interrupt_cascade_register(dsp_irq + i);
```
Comment (paulstelian97) [inline src/drivers/imx/interrupt-irqstr.c:38]: Might want to refactor the body of this function into a separate irqstr_init(), and platform_interrupt_init just calls irqstr_init().

File: src/include/sof/drivers/irqstr.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+#ifndef __SOF_DRIVERS_IRQSTR_H__
+#define __SOF_DRIVERS_IRQSTR_H__
+
+void irqstr_write(uint32_t reg, uint32_t value);
+
+void irqstr_enable_hw(void);
+void irqstr_disable_hw(void);
+
+#endif
```
Comment (paulstelian97) [inline src/include/sof/drivers/irqstr.h:15]: **Do not break build (headers added after files that use them)**

File: src/include/sof/drivers/interrupt.h [hunk 4]
```diff
@@ -96,6 +96,8 @@ struct irq_cascade_desc {
 							  */
 };
 
+#define DSP_IRQ_MAX 8
```
Comment (paulstelian97) [inline src/include/sof/drivers/interrupt.h:99]: **Do not break build (macro added after use)**

**Review by paulstelian97 — CHANGES_REQUESTED** [review]
To avoid breaking bisect, you should combine these commits into one (adding the header and macro can be separate but it should be _before_ the first time you use the header, not after)

**Comment by dbaluta** [comment]
@TangleZ 

Hi Peng,

Thanks a lot for taking care of this. After having a better look at the code I think we can start with separate interrupt drivers for 
platforms with irqsteer and platforms without irqsteer.

So, lets move the existing code from ``interrupt.c`` to ``interrupt-irqsteer.c``
```
$ git mv interrupt.c interrupt-irqsteer.c
```
then update CMakeLists.txt

```
diff --git a/src/drivers/imx/CMakeLists.txt b/src/drivers/imx/CMakeLists.txt
index 94e3c11ec..cc221a36c 100644
--- a/src/drivers/imx/CMakeLists.txt
+++ b/src/drivers/imx/CMakeLists.txt
@@ -3,7 +3,6 @@
 add_local_sources(sof
        esai.c
        sai.c
-       interrupt.c
        ipc.c
        timer.c
        sdma.c
@@ -12,3 +11,7 @@ add_local_sources(sof
 if(CONFIG_IMX_EDMA)
        add_local_sources(sof edma.c)
 endif()
+
+if(CONFIG_IMX_INTERRUPT_IRQSTEER)
+       add_local_sources(sof interrupt-irqsteer.c)
+endif()

```

Then select by default IMX_INTERRUPT_IRQSTEER by default for imx8/8x/8m

```
--- a/src/platform/Kconfig
+++ b/src/platform/Kconfig
@@ -179,6 +179,7 @@ config IMX8
        select IMX
        select IMX_EDMA
        select SCHEDULE_DMA_MULTI_CHANNEL
+       select IMX_INTERRUPT_IRQSTEER
        help
          Select if your target platform is imx8-compatible
 
@@ -194,6 +195,7 @@ config IMX8X
        select IMX
        select IMX_EDMA
        select SCHEDULE_DMA_MULTI_CHANNEL
+       select IMX_INTERRUPT_IRQSTEER
        help
          Select if your target platform is imx8x-compatible
 
@@ -208,6 +210,7 @@ config IMX8M
        select WAITI_DELAY
        select IMX
        select SCHEDULE_DMA_MULTI_CHANNEL
+       select IMX_INTERRUPT_IRQSTEER
        help
          Select if your target platform is imx8m-compatible

```

After this patch is in you can add your interrupt support for imx8ulp.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by dbaluta** [comment]
@plbossart @lgirdwood can you please add @TangleZ to the list of SOF developers so that CI can run?

**Comment by paulstelian97** [comment]
> Can one of the admins verify this patch?

test this please

**Comment by TangleZ** [comment]
Updated the patch.

**Comment by dbaluta** [comment]
Thanks @TangleZ  this looks good now! One minor thing is that you also need to add a Kconfig entry for CONFIG_IMX_INTERRUPT_IRQSTEER.

 But you will need to wait that  my PR https://github.com/thesofproject/sof/pull/4013 gets merged first. Just look add #4013 how the Kconfig entry should be added.

**Comment by lgirdwood** [comment]
please test this

**Comment by lgirdwood** [comment]
@TangleZ please look out for an invite email.

**Comment by TangleZ** [comment]
@lgirdwood How to test the patch? I test the patch by run ./scripts/xtensa-build-all.sh imx8, it's ok?

**Comment by dbaluta** [comment]
@TangleZ 'please test this' is a message to CI not for you.

Now that https://github.com/thesofproject/sof/pull/4013/commits/f2a969a7f1781cde53e7220eda1645a3227b480b is merged can you please also update Kconfig to include your new config option? CONFIG_INTERRUPT_IRQSTEER?

**Review by paulstelian97 — COMMENTED** [review]
Does this build correctly for 8ULP?

**Comment by TangleZ** [comment]
@paulstelian97 This patch is not include imx8ulp part, just for preparing.

File: src/platform/Kconfig [hunk 5]
```diff
@@ -256,6 +259,13 @@ config IMX
 	  This has to be selected for every i.MX NXP platform.
 	  It enables NXP platforms-specific features.
 
+config IMX_INTERRUPT_IRQSTEER
+	bool
+	default n
+	help
+	  This has to be selected for i.MX NXP platform that have irqsteer.
+	  It enables NXP platforms-specific features.
+
```
Comment (dbaluta) [inline src/platform/Kconfig:268]: @TangleZ please move this config definition to src/drivers/imx/Kconfig.

Other than that, looks good to me.

**Review by paulstelian97 — APPROVED** [review]
Approving on my end, since the only change I see needed is already mentioned by Daniel.

**Review by dbaluta — APPROVED** [review]

**Comment by lgirdwood** [comment]
@zrombel can you comment on the DMIC failure - it looks unrelated to this PR.

File: src/drivers/imx/CMakeLists.txt [hunk 6]
```diff
@@ -3,7 +3,6 @@
 add_local_sources(sof
 	esai.c
 	sai.c
-	interrupt.c
```
Comment (lgirdwood) [inline src/drivers/imx/CMakeLists.txt:6]: Do we need to delete this or is it still used ?

**Review by lgirdwood — APPROVED** [review]

File: src/drivers/imx/CMakeLists.txt (hunk 6 above)
Comment (paulstelian97) [reply src/drivers/imx/CMakeLists.txt:6]: The file was renamed and moved to be conditional. Currently all i.MX platforms that work have irqsteer enabled so it doesn't matter. @TangleZ said in internal mailing lists that he will send a second pull request to actually add the support proper for 8ULP from this point of view.

I expect a future pull request to add a new interrupt.c file which has some simple, empty functions to satisfy the linker, given that 8ULP only has arch interrupts and the arch_ functions suffice for our use case.

**Comment by zrombel** [comment]
This failure was unrelated. All green after rerun.

**Pull request merged**